Warn users about ns label warnings

Jaanki opened the pull request:
K8S1.23/OCP 4.11 and above throws warnings if the ns is not
`PodSecurity` labelled. This affects `diagnose kube-proxy-mode`
and `diagnose firewall` subcommands.

This PR adds a warning about these warnings so that the user
is not surprised when they pop up in the logs while running
these commands.

Partially Closes: #119

Depends on https://github.com/submariner-io/subctl/pull/126

Sample output when the `namespace` has not been specified by the user:
```
./cmd/bin/subctl diagnose firewall nat-discovery output/kubeconfigs/kind-config-cluster1 output/kubeconfigs/kind-config-cluster2
 ✓ Checking if nat-discovery port is opened on the gateway node of cluster "cluster1" 
 ✓ nat-discovery port is allowed in the cluster "cluster1"
```

Output when the namespace has been specified by the user:
```
./cmd/bin/subctl diagnose firewall nat-discovery output/kubeconfigs/kind-config-cluster1 output/kubeconfigs/kind-config-cluster2 --namespace user_ns
 ⚠ Kubernetes 1.23/OpenShift Container Platform 4.11 and above expects namespaces to have security labels as required by `PodSecurity Admission controller`. Otherwise, the related warnings would be logged. To avoid such warnings, make sure the namespace specified has below labels:
 ⚠ pod-security.kubernetes.io/enforce: privileged
 ⚠ pod-security.kubernetes.io/audit: privileged
 ⚠ pod-security.kubernetes.io/warn: privileged
```


Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

skitt reviewed (changes requested):
Please check whether the requested namespace has the appropriate labels, and only warn if it doesn’t.

Jaanki committed:
Add warning about missing labels

PodSecurity Admission controllers (k8s 1.23 >=/OCP 4.11 >=)
require few labels added to the namespace. This PR checks the
namespace for the labels. If the label is missing or is incorrect,
a warning is thrown to the user about the missing labels and
relevant warnigs logged by the k8s.

Partially Closes: #119

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

Warn users about ns label warnings

K8S1.23/OCP 4.11 and above throws warnings if the ns is not
`PodSecurity` labelled. This affects `diagnose kube-proxy-mode`
and `diagnose firewall` subcommands.

This PR adds a warning about these warnings so that the user
is not surprised when they pop up in the logs while running
these commands.

Partially Closes: #119

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

skitt commented on internal/pods/schedule.go:
```diff
@@ -270,3 +285,35 @@ func addNodeSelectorTerm(nodeSelTerms []v1.NodeSelectorTerm, label string,
+
+	missingLabels := map[string]string{}
+
+	for key := range expectedLabels {
+		val, found := nsLabels[key]
+		if !found || val != "privileged" {
+			missingLabels[key] = "privileged"
+		}
+	}
```
Instead of hard-coding the expected values here, use those from the map:

```suggestion
	for key, expectedLabel := range expectedLabels {
		actualLabel, found := nsLabels[key]
		if !found || actualLabel != expectedLabel {
			missingLabels[key] = expectedLabel
		}
	}
```

skitt commented on internal/pods/schedule.go:
```diff
@@ -270,3 +285,35 @@ func addNodeSelectorTerm(nodeSelTerms []v1.NodeSelectorTerm, label string,
+
+func warnAbout(missingLabels map[string]string) {
+	status := cli.NewReporter()
+	status.Warning("Kubernetes 1.23/OpenShift Container Platform 4.11 and above expects namespaces to have security labels as required by" +
+		" `PodSecurity Admission controller`. Otherwise, the related warnings would be logged. To avoid such warnings, make sure the" +
+		" namespace specified has below labels:")
```
For upstream Submariner, reduce the message length by only mentioning K8s, and indicate that the warnings are innocuous:

```suggestion
	status.Warning("Kubernetes 1.23 and above, the Pod Security admission controller expects namespaces to have security labels. Without these, you will see warnings in subctl's output. subctl should work fine, but you can avoid the warnings and ensure correct behavior by adding these labels to the namespace:")
```

(unwrapped). While we’re at it, please specify the namespace in the output — the user just specified it but it’s always best to have fully actionable error messages.

mkolesnik commented on internal/pods/schedule.go:
```diff
@@ -78,6 +79,13 @@ func ScheduleAndAwaitCompletion(config *Config) (string, error) {
 		config.Namespace = constants.OperatorNamespace
 	}
 
+	ns, err := config.ClientSet.CoreV1().Namespaces().Get(context.TODO(), config.Namespace, metav1.GetOptions{})
+	if err != nil {
+		return "", errors.Wrap(err, fmt.Sprintf("error fetching %s namespace", config.Namespace))
+	}
+
+	checkNSLabels(ns)
+
```
```suggestion
	if err := checkNSLabels(config); err != nil {
		return "", err
	}

```

mkolesnik commented on internal/pods/schedule.go:
```diff
@@ -271,3 +286,37 @@ func addNodeSelectorTerm(nodeSelTerms []v1.NodeSelectorTerm, label string,
 	}})
 }
+
+func checkNSLabels(ns *v1.Namespace) {
```
```suggestion
func checkNSLabels(config *Config) error {
	ns, err := config.ClientSet.CoreV1().Namespaces().Get(context.TODO(), config.Namespace, metav1.GetOptions{})
	if err != nil {
		return errors.Wrap(err, fmt.Sprintf("error fetching %s namespace", config.Namespace))
	}

```

Less repetition, plus namespace object isn't used outside so its OK to fetch it just for this function

mkolesnik commented on internal/pods/schedule.go:
```diff
@@ -101,6 +109,13 @@ func Schedule(config *Config) (*Scheduled, error) {
 		config.Namespace = constants.OperatorNamespace
 	}
 
+	ns, err := config.ClientSet.CoreV1().Namespaces().Get(context.TODO(), config.Namespace, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("error fetching %s namespace", config.Namespace))
+	}
+
+	checkNSLabels(ns)
+
```
```suggestion
	if err := checkNSLabels(config); err != nil {
		return "", err
	}

```

mkolesnik commented on internal/pods/schedule.go:
```diff
@@ -271,3 +286,37 @@ func addNodeSelectorTerm(nodeSelTerms []v1.NodeSelectorTerm, label string,
+
+func warnAbout(missingLabels map[string]string, namespace string) {
+	status := cli.NewReporter()
+	status.Warning("Kubernetes 1.23 and above, the Pod Security admission controller expects namespaces to have security labels. Without"+
```
```suggestion
	status.Warning("Starting with Kubernetes 1.23, the Pod Security admission controller expects namespaces to have security labels. Without"+
```

mkolesnik reviewed (commented):
A few suggestions
